### PR TITLE
Review/jm initial

### PR DESCRIPTION
**Pull request by fivetran-joemarkiewicz**
This PR is serving as my review of PR #40. Please see comments below for review notes and suggestions. This PR does not necessarily need to be merged. Please comment in the comments once you have addressed them in your branch.

### CODE REVIEW

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-avinash see my review notes below. Some themes to highlight here include:

- We should make the staging models incremental for history mode.
- Consolidating the README instructions for history mode into a single section for easier use.
- Other specific notes you can find in my comments.

In addition to my comments below I also was wondering if we could have the hierarchy of the models be something like the following:

```sh
models
└ history
    └ stg_salesforce__contact_history.sql
    └ etc.
└ standard
    └ stg_salesforce__contact_.sql
    └ etc.
    └ tmp
       └ stg_salesforce__contact_tmp.sql
       └ etc.
```
This way it is clear what the separation of the models are. Thoughts?

Lastly, I wanted to get confirmation on the use cases we are trying to cover here. Can you confirm that this update should allow the following use cases, or am I missing one or two or if one of them is not accurate?
- A customer only using the standard models and not history
- A customer using the standard models with a standard connector and the history models with a history connector
- A customer using the standard models with the history connector and the history models with a history connector

Let me know if you have any questions!

File: models/history/stg_salesforce__contact_history.sql [hunk 1]
```diff
@@ -55,13 +71,14 @@ final as (
...
+        {{ dbt_utils.generate_surrogate_key(['id', '_fivetran_start']) }} as history_unique_key
+
         {{ fivetran_utils.fill_pass_through_columns('salesforce__contact_history_pass_through_columns') }}
         
     from fields
 )
 
 select *
-from final
-where not coalesce(is_deleted, false)
```
Comment (fivetran-joemarkiewicz) [inline models/history/stg_salesforce__contact_history.sql:67]: Should this still be here for history mode models?

File: models/history/stg_salesforce__contact_history.sql [hunk 2]
```diff
@@ -1,17 +1,33 @@
...
+{{ config(
+        enabled=var('salesforce__contact_history_enabled', False),
+        materialized='incremental',
+        unique_key='history_unique_key',
+        incremental_strategy='insert_overwrite' if target.type in ('bigquery', 'spark', 'databricks') else 'delete+insert',
+        partition_by={"field": "_fivetran_date", "data_type": "date"} if target.type not in ('spark','databricks') else ['_fivetran_date'],
+        file_format='parquet',
+        on_schema_change='fail'
+    ) 
+}}
```
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/history/stg_salesforce__contact_history.sql:10]: Additional, if we add this we should include a section in the README that highlights the nuances of incremental models. Like we do [here](https://github.com/fivetran/dbt_fivetran_log/tree/main#database-incremental-strategies).

**Comment by fivetran-avinash** [comment]
> @fivetran-avinash see my review notes below. Some themes to highlight here include:
> 
> * We should make the staging models incremental for history mode.
> * Consolidating the README instructions for history mode into a single section for easier use.
> * Other specific notes you can find in my comments.
> 
> In addition to my comments below I also was wondering if we could have the hierarchy of the models be something like the following:
> 
> ```shell
> models
> └ history
>     └ stg_salesforce__contact_history.sql
>     └ etc.
> └ standard
>     └ stg_salesforce__contact_.sql
>     └ etc.
>     └ tmp
>        └ stg_salesforce__contact_tmp.sql
>        └ etc.
> ```
> 
> This way it is clear what the separation of the models are. Thoughts?
> 
> Lastly, I wanted to get confirmation on the use cases we are trying to cover here. Can you confirm that this update should allow the following use cases, or am I missing one or two or if one of them is not accurate?
> 
> * A customer only using the standard models and not history
> * A customer using the standard models with a standard connector and the history models with a history connector
> * A customer using the standard models with the history connector and the history models with a history connector
> 
> Let me know if you have any questions!
> 
> Edit: I also have not reviewed the CHANGELOG as I imagine there may be some changes following this review.

@fivetran-joemarkiewicz 
Wonder if the folders should just be named after the schemas for simplicity? I.e. salesforce vs salesforce_history.

When you say a history connector, you mean history mode is enabled and two separate schemas exist for salesforce and salesforce history? I believe those are the only three cases we've discussed!

**Comment by fivetran-joemarkiewicz** [comment]
@fivetran-avinash I have been reviewing the transform updates and it just occurred to me that the only history models that are being used in downstream transformations are `contact`, `account`, and `opportunity`. We are not doing joins with any other models or leveraging them in anyway. With that, I am wondering what value we are bringing to the package by including the history versions of all the other sources in this update?

I could see a case where they do add value, but I am curious what your thoughts are on why they were included.

File: models/history/stg_salesforce__contact_history.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/history/stg_salesforce__contact_history.sql:10]: One note here is use your due diligence with the unique key

File: models/history/stg_salesforce__contact_history.sql [hunk 3]
```diff
@@ -1,17 +1,33 @@
...
     select * 
-    from {{ ref('stg_salesforce__contact_history_tmp') }}
+    from {{ source('salesforce_history','contact') }}
+    {% if is_incremental() %}
+    where date(_fivetran_start) >= date(_dbt_max_partition)
+    {% else %}
+    {% if var('contact_first_date_var',[]) %}
+    where _fivetran_start >= '{{ var('account_first_date_var') }}'
+    {% endif %}
+    {% endif %}
```
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/history/stg_salesforce__contact_history.sql]: Look at commit where this is changed to leverage both global and specific date filters

File: models/history/stg_salesforce__contact_history.sql [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
 ),
 
 fields as (
 
     select
         {{
             fivetran_utils.fill_staging_columns(
-                source_columns=adapter.get_columns_in_relation(ref('stg_salesforce__contact_history_tmp')),
+                source_columns=adapter.get_columns_in_relation(source('salesforce_history','contact')),
```
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply models/history/stg_salesforce__contact_history.sql:30]: We may also be able to remove the need for the get columns macro, but let's double check that if we are wanting to rename any fields.

File: README.md [hunk 5]
```diff
@@ -187,7 +187,7 @@ Include the following within your `dbt_project.yml` file:
 ```
 
 ### Adding Passthrough Columns
-This package includes all source columns defined in the `generate_columns.sql` macro.  You can add more columns using our passthrough column variables. These variables allow for the passthrough fields to be aliased (`alias`) and casted (`transform_sql`) if desired, but not required. Datatype casting is configured via a sql snippet within the `transform_sql` key. You may add the desired sql while omitting the `as field_name` at the end and your custom pass-though fields will be casted accordingly. Use the below format for declaring the respective pass-through variables:
+This package includes all source columns defined in the `generate_columns.sql` macro.  You can add more columns using our passthrough column variables. These variables allow for the passthrough fields to be aliased (`alias`) and casted (`transform_sql`) if desired, but not required. Datatype casting is configured via a sql snippet within the `transform_sql` key. You may add the desired sql while omitting the `as field_name` at the end and your custom pass-though fields will be casted accordingly. Use the below format for declaring the respective pass-through variables:
```
[the remark this answers is not in the document]
Comment (fivetran-joemarkiewicz) [reply README.md:190]: We may not even use passthrough for history anymore

File: .buildkite/scripts/run_models.sh [hunk 6]
```diff
@@ -23,4 +23,6 @@ dbt run --vars '{salesforce__user_role_enabled: false}' --target "$db" --full-re
 dbt test --target "$db"
 dbt run --vars '{salesforce__account_history_enabled: true, salesforce__contact_history_enabled: true, salesforce__event_history_enabled: true, salesforce__lead_history_enabled: true, salesforce__opportunity_history_enabled: true, salesforce__opportunity_line_item_history_enabled: true, salesforce__product_2_history_enabled: true, salesforce__task_history_enabled: true, salesforce__user_history_enabled: true, salesforce__user_role_history_enabled: true}' --target "$db" --full-refresh
 dbt test --target "$db"
+dbt run --vars '{salesforce__account_history_enabled: true, salesforce__contact_history_enabled: true, salesforce__event_history_enabled: true, salesforce__lead_history_enabled: true, salesforce__opportunity_history_enabled: true, salesforce__opportunity_line_item_history_enabled: true, salesforce__product_2_history_enabled: true, salesforce__task_history_enabled: true, salesforce__user_history_enabled: true, salesforce__user_role_history_enabled: true}' --target "$db"
+dbt test --target "$db"
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply .buildkite/scripts/run_models.sh:27]: Updated, and removed the variables not being used.

File: models/history/stg_salesforce__product_2_history.sql [hunk 7]
```diff
@@ -52,9 +52,9 @@ final as (
         revenue_installment_period,
         revenue_schedule_type
 
-    from fields
+        {{ fivetran_utils.fill_pass_through_columns('salesforce__product_2_history_pass_through_columns') }}
 
-    {{ fivetran_utils.add_pass_through_columns(columns, var('salesforce__product_2_history_pass_through_columns')) }}
+    from fields
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/history/stg_salesforce__product_2_history.sql:57]: This is being removed.

File: models/history/stg_salesforce__opportunity_line_item_history.sql [hunk 8]
```diff
@@ -52,9 +52,9 @@ final as (
         total_price,
         unit_price
 
-    from fields
+        {{ fivetran_utils.fill_pass_through_columns('salesforce__opportunity_line_item_history_pass_through_columns') }}
 
-    {{ fivetran_utils.add_pass_through_columns(columns, var('salesforce__opportunity_line_item_history_pass_through_columns')) }}
+    from fields
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/history/stg_salesforce__opportunity_line_item_history.sql:57]: This model is being removed from this release.

File: models/history/stg_salesforce__contact_history.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/history/stg_salesforce__contact_history.sql:30]: Gone ahead and removed the get columns macro per our discussion as we are now bringing in all synced fields.

File: models/history/stg_salesforce__contact_history.sql (hunk 1 above)
Comment (fivetran-avinash) [reply models/history/stg_salesforce__contact_history.sql:67]: Removed.

File: models/src_salesforce.yml [hunk 9]
```diff
@@ -9,8 +9,6 @@ sources:
     freshness:
       warn_after: {count: 24, period: hour}
       error_after: {count: 48, period: hour}
-    config:
-      enabled: true
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/src_salesforce.yml:13]: Removed

File: models/history/stg_salesforce__contact_history.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/history/stg_salesforce__contact_history.sql]: This logic is now updated and verified to work in a dbt run.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:190]: We won't, so no action needed here I believe?

**Comment by fivetran-avinash** [comment]
> This PR is serving as my review of PR #40. Please see comments below for review notes and suggestions. This PR does not necessarily need to be merged. Please comment in the comments once you have addressed them in your branch.

PR recommendations acknowledged and applied! [See here for further notes. ](https://github.com/fivetran/dbt_salesforce_source/pull/40#issuecomment-1779350551)

**Comment by fivetran-avinash** [comment]
> This PR is serving as my review of PR #40. Please see comments below for review notes and suggestions. This PR does not necessarily need to be merged. Please comment in the comments once you have addressed them in your branch.

@fivetran-joemarkiewicz  PR recommendations acknowledged and applied! [See here for further notes and additional lines of investigation. ](https://github.com/fivetran/dbt_salesforce_source/pull/40#issuecomment-1779350551)

**Pull request closed without merging**